File: app/src/main/java/com/clockmods/sdk/clock/ClockState.java
package com.clockmods.sdk.clock;

import java.util.Calendar;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Locale;
import java.util.TimeZone;

/** Immutable clock data. It contains no layout, paint, or theme decisions. */
public final class ClockState {
    public enum SecondHandMotion { OFF, TICK, SWEEP }

    private final long timeMillis;
    private final TimeZone timeZone;
    private final Locale locale;
    private final boolean use24Hour;
    private final boolean showSeconds;
    private final SecondHandMotion secondHandMotion;
    private final String dateText;
    private final String timeZoneText;
    private final String weatherText;
    private final String statusText;
    private final List<WorldClockEntry> worldClocks;
    private final float timeScale;
    private final float dateScale;
    private final float supportingScale;

    private ClockState(Builder builder) {
        timeMillis = builder.timeMillis;
        timeZone = (TimeZone) builder.timeZone.clone();
        locale = builder.locale;
        use24Hour = builder.use24Hour;
        showSeconds = builder.showSeconds;
        secondHandMotion = builder.secondHandMotion;
        dateText = clean(builder.dateText);
        timeZoneText = clean(builder.timeZoneText);
        weatherText = clean(builder.weatherText);
        statusText = clean(builder.statusText);
        worldClocks = Collections.unmodifiableList(new ArrayList<WorldClockEntry>(
                builder.worldClocks));
        timeScale = positiveScale(builder.timeScale);
        dateScale = positiveScale(builder.dateScale);
        supportingScale = positiveScale(builder.supportingScale);
    }

    public static Builder builder(long timeMillis) {
        return new Builder(timeMillis);
    }

    public long getTimeMillis() { return timeMillis; }
    public TimeZone getTimeZone() { return (TimeZone) timeZone.clone(); }
    public Locale getLocale() { return locale; }
    public boolean isUse24Hour() { return use24Hour; }
    public boolean isShowSeconds() { return showSeconds; }
    public SecondHandMotion getSecondHandMotion() { return secondHandMotion; }
    public String getDateText() { return dateText; }
    public String getTimeZoneText() { return timeZoneText; }
    public String getWeatherText() { return weatherText; }
    public String getStatusText() { return statusText; }
    public List<WorldClockEntry> getWorldClocks() { return worldClocks; }
    public float getTimeScale() { return timeScale; }
    public float getDateScale() { return dateScale; }
    public float getSupportingScale() { return supportingScale; }

    public Calendar newCalendar() {
        Calendar calendar = Calendar.getInstance(timeZone, locale);
        calendar.setTimeInMillis(timeMillis);
        return calendar;
    }

    private static String clean(String value) {
        return value == null ? "" : value.trim();
    }

    public static final class Builder {
        private final long timeMillis;
        private TimeZone timeZone = TimeZone.getDefault();
        private Locale locale = Locale.getDefault();
        private boolean use24Hour = true;
        private boolean showSeconds = true;
        private SecondHandMotion secondHandMotion = SecondHandMotion.TICK;
        private String dateText = "";
        private String timeZoneText = "";
        private String weatherText = "";
        private String statusText = "";
        private List<WorldClockEntry> worldClocks = Collections.emptyList();
        private float timeScale = 1f;
        private float dateScale = 1f;
        private float supportingScale = 1f;

        private Builder(long timeMillis) {
            this.timeMillis = timeMillis;
        }

        public Builder timeZone(TimeZone value) {
            if (value == null) throw new IllegalArgumentException("timeZone must not be null");
            timeZone = (TimeZone) value.clone();
            return this;
        }

        public Builder locale(Locale value) {
            if (value == null) throw new IllegalArgumentException("locale must not be null");
            locale = value;
            return this;
        }

        public Builder use24Hour(boolean value) { use24Hour = value; return this; }
        public Builder showSeconds(boolean value) { showSeconds = value; return this; }

        public Builder secondHandMotion(SecondHandMotion value) {
            if (value == null) {
                throw new IllegalArgumentException("secondHandMotion must not be null");
            }
            secondHandMotion = value;
            return this;
        }

        public Builder dateText(String value) { dateText = value; return this; }
        public Builder timeZoneText(String value) { timeZoneText = value; return this; }
        public Builder weatherText(String value) { weatherText = value; return this; }
        public Builder statusText(String value) { statusText = value; return this; }

        public Builder worldClocks(List<WorldClockEntry> value) {
            worldClocks = value == null ? Collections.<WorldClockEntry>emptyList()
                    : new ArrayList<WorldClockEntry>(value);
            return this;
        }

        public Builder timeScale(float value) { timeScale = value; return this; }
        public Builder dateScale(float value) { dateScale = value; return this; }
        public Builder supportingScale(float value) { supportingScale = value; return this; }

        public ClockState build() {
            return new ClockState(this);
        }
    }

    private static float positiveScale(float value) {
        return Float.isNaN(value) || Float.isInfinite(value) || value <= 0f ? 1f : value;
    }
}

File: app/src/main/java/com/clockmods/sdk/clock/ClockState.kt
package com.clockmods.sdk.clock

import java.util.Calendar
import java.util.Collections
import java.util.Locale
import java.util.TimeZone

/** Immutable clock data. It contains no layout, paint, or theme decisions. */
class ClockState private constructor(
    timeMillis: Long,
    timeZone: TimeZone,
    locale: Locale,
    use24Hour: Boolean,
    showSeconds: Boolean,
    blinkColon: Boolean,
    smallSeconds: Boolean,
    portraitStacked: Boolean,
    dateLunarDualLine: Boolean,
    secondHandMotion: SecondHandMotion,
    dateText: String?,
    timeZoneText: String?,
    weatherText: String?,
    statusText: String?,
    worldClocks: List<WorldClockEntry>,
    timeScale: Float,
    dateScale: Float,
    supportingScale: Float,
    timeTransition: TimeTransition,
    timeTransitionProgress: Float,
) {
    enum class SecondHandMotion { OFF, TICK, SWEEP }
    enum class TimeTransition { FADE, SLIDE_UP, SLIDE_DOWN, SCALE, FLIP }

    private val timeMillis = timeMillis
    private val timeZone = timeZone.clone() as TimeZone
    private val locale = locale
    private val use24Hour = use24Hour
    private val showSeconds = showSeconds
    private val blinkColon = blinkColon
    private val smallSeconds = smallSeconds
    private val portraitStacked = portraitStacked
    private val dateLunarDualLine = dateLunarDualLine
    private val secondHandMotion = secondHandMotion
    private val dateText = clean(dateText)
    private val timeZoneText = clean(timeZoneText)
    private val weatherText = clean(weatherText)
    private val statusText = clean(statusText)
    private val worldClocks = Collections.unmodifiableList(worldClocks.toList())
    private val timeScale = positiveScale(timeScale)
    private val dateScale = positiveScale(dateScale)
    private val supportingScale = positiveScale(supportingScale)
    private val timeTransition = timeTransition
    private val timeTransitionProgress = transitionProgress(timeTransitionProgress)

    fun getTimeMillis() = timeMillis
    fun getTimeZone() = timeZone.clone() as TimeZone
    fun getLocale() = locale
    fun isUse24Hour() = use24Hour
    fun isShowSeconds() = showSeconds
    fun isBlinkColon() = blinkColon
    fun isSmallSeconds() = smallSeconds
    fun isPortraitStacked() = portraitStacked
    fun isDateLunarDualLine() = dateLunarDualLine
    fun getSecondHandMotion() = secondHandMotion
    fun getDateText() = dateText
    fun getTimeZoneText() = timeZoneText
    fun getWeatherText() = weatherText
    fun getStatusText() = statusText
    fun getWorldClocks() = worldClocks
    fun getTimeScale() = timeScale
    fun getDateScale() = dateScale
    fun getSupportingScale() = supportingScale
    fun getTimeTransition() = timeTransition
    fun getTimeTransitionProgress() = timeTransitionProgress

    fun newCalendar(): Calendar = Calendar.getInstance(timeZone, locale).also {
        it.timeInMillis = timeMillis
    }

    class Builder private constructor(private val timeMillis: Long) {
        private var timeZone: TimeZone = TimeZone.getDefault()
        private var locale: Locale = Locale.getDefault()
        private var use24Hour = true
        private var showSeconds = true
        private var blinkColon = false
        private var smallSeconds = false
        private var portraitStacked = false
        private var dateLunarDualLine = false
        private var secondHandMotion = SecondHandMotion.TICK
        private var dateText = ""
        private var timeZoneText = ""
        private var weatherText = ""
        private var statusText = ""
        private var worldClocks: List<WorldClockEntry> = emptyList()
        private var timeScale = 1f
        private var dateScale = 1f
        private var supportingScale = 1f
        private var timeTransition = TimeTransition.FADE
        private var timeTransitionProgress = 1f

        fun timeZone(value: TimeZone?): Builder = apply {
            requireNotNull(value) { "timeZone must not be null" }
            timeZone = value.clone() as TimeZone
        }

        fun locale(value: Locale?): Builder = apply {
            locale = requireNotNull(value) { "locale must not be null" }
        }

        fun use24Hour(value: Boolean) = apply { use24Hour = value }
        fun showSeconds(value: Boolean) = apply { showSeconds = value }
        fun blinkColon(value: Boolean) = apply { blinkColon = value }
        fun smallSeconds(value: Boolean) = apply { smallSeconds = value }
        fun portraitStacked(value: Boolean) = apply { portraitStacked = value }
        fun dateLunarDualLine(value: Boolean) = apply { dateLunarDualLine = value }

        fun secondHandMotion(value: SecondHandMotion?): Builder = apply {
            secondHandMotion = requireNotNull(value) { "secondHandMotion must not be null" }
        }

        fun dateText(value: String?) = apply { dateText = value.orEmpty() }
        fun timeZoneText(value: String?) = apply { timeZoneText = value.orEmpty() }
        fun weatherText(value: String?) = apply { weatherText = value.orEmpty() }
        fun statusText(value: String?) = apply { statusText = value.orEmpty() }

        fun worldClocks(value: List<WorldClockEntry>?): Builder = apply {
            worldClocks = value?.toList() ?: emptyList()
        }

        fun timeScale(value: Float) = apply { timeScale = value }
        fun dateScale(value: Float) = apply { dateScale = value }
        fun supportingScale(value: Float) = apply { supportingScale = value }

        fun timeTransition(value: TimeTransition?) = apply {
            timeTransition = requireNotNull(value) { "timeTransition must not be null" }
        }

        fun timeTransitionProgress(value: Float) = apply { timeTransitionProgress = value }

        fun build() = ClockState(
            timeMillis,
            timeZone,
            locale,
            use24Hour,
            showSeconds,
            blinkColon,
            smallSeconds,
            portraitStacked,
            dateLunarDualLine,
            secondHandMotion,
            dateText,
            timeZoneText,
            weatherText,
            statusText,
            worldClocks,
            timeScale,
            dateScale,
            supportingScale,
            timeTransition,
            timeTransitionProgress,
        )

        internal companion object {
            @JvmSynthetic
            fun create(timeMillis: Long) = Builder(timeMillis)
        }
    }

    companion object {
        @JvmStatic
        fun builder(timeMillis: Long) = Builder.create(timeMillis)

        private fun clean(value: String?) = value?.trim().orEmpty()
        private fun positiveScale(value: Float) =
            if (value.isNaN() || value.isInfinite() || value <= 0f) 1f else value
        private fun transitionProgress(value: Float) =
            if (value.isNaN()) 1f else value.coerceIn(0f, 1f)
    }
}
